src/States.cpp

```diff
@@ -0,0 +1,291 @@
+
+   // Empty constructor
+   States::States() : m_refFrame(0) {
+    m_states = {};
+    m_ephemTimes = {};
+  }
+
+
+  States::States(const std::vector<double>& ephemTimes, const std::vector<ale::Vec3d>& positions,
+                 int refFrame) :
+    m_ephemTimes(ephemTimes), m_refFrame(refFrame) {
+    // Construct State vector from position and velocity vectors
+
+    if (positions.size() != ephemTimes.size()) {
+      throw std::invalid_argument("Length of times must match number of positions");
+    }
+
+    ale::Vec3d velocities = {0.0, 0.0, 0.0};
+    for (ale::Vec3d position : positions) {
+      m_states.push_back(ale::State(position, velocities));
+    }
+  }
+
+
+  States::States(const std::vector<double>& ephemTimes, const std::vector<ale::Vec3d>& positions,
+                 const std::vector<ale::Vec3d>& velocities, int refFrame) :
+    m_ephemTimes(ephemTimes), m_refFrame(refFrame) {
+
+    if ((positions.size() != ephemTimes.size())||(ephemTimes.size() != velocities.size())) {
+      throw std::invalid_argument("Length of times must match number of positions and velocities.");
+    }
+
+    for (int i=0; i < positions.size() ;i++) {
+      m_states.push_back(ale::State(positions[i], velocities[i]));
+    }
+  }
+
+
+  States::States(const std::vector<double>& ephemTimes, const std::vector<ale::State>& states,
+                 int refFrame) :
+  m_ephemTimes(ephemTimes), m_states(states), m_refFrame(refFrame) {
+    if (states.size() != ephemTimes.size()) {
+      throw std::invalid_argument("Length of times must match number of states.");
+    }
+  }
+
+  // Default Destructor
+  States::~States() {}
+
+  // Getters
+  std::vector<ale::State> States::getStates() const {
+    return m_states;
+  }
+
+  std::vector<ale::Vec3d> States::getPositions() const {
+    // extract positions from state vector
+    std::vector<ale::Vec3d> positions;
+
+    for(ale::State state : m_states) {
+        positions.push_back(state.position);
+    }
+    return positions;
+  }
+
+
+  std::vector<ale::Vec3d> States::getVelocities() const {
+    // extract velocities from state vector
+    std::vector<ale::Vec3d> velocities;
+
+    for(ale::State state : m_states) {
+        velocities.push_back(state.velocity);
+    }
+    return velocities;
+  }
+
+
+  std::vector<double> States::getTimes() const {
+    return m_ephemTimes;
+  }
+
+
+  int States::getReferenceFrame() const {
+    return m_refFrame;
+  }
+
+
+  bool States::hasVelocity() const {
+    std::vector<ale::Vec3d> velocities = getVelocities();
+    bool allZero = std::all_of(velocities.begin(), velocities.end(), [](ale::Vec3d vec)
+                               { return vec.x==0.0 && vec.y==0.0 && vec.z==0.0; });
+    return !allZero;
+  }
+
+
+  ale::State States::getState(double time, ale::interpolation interp) const {
+    std::vector<double> xs, ys, zs, vxs, vys, vzs;
+    int lowerBound = ale::interpolationIndex(m_ephemTimes, time);
+    std::vector<double> interpTimes = {};
```

Comment:
Move the definitions of xs, ys, zs, vxs, vys, vzs, and interpTimes down to just before they are filled. It'd confusing to have them separated by the interpolation range logic.